### ISSUE TITLE
Incompatible with webpack@4.0.0-alpha.1 (the @next version as of now)

### ISSUE THREAD

**Issue by sompylasar** [amireh/happypack#200]
```
TypeError: Cannot read property 'length' of undefined
at resolveLoader (/__CENSORED__/node_modules/happypack/lib/WebpackUtils.js:138:17)
at /__CENSORED__/node_modules/happypack/lib/WebpackUtils.js:126:7
at /__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:713:13
at async.forEachOf.async.eachOf (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:233:13)
at _parallel (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:712:9)
at Object.async.parallel (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:726:9)
at Object.exports.resolveLoaders (/__CENSORED__/node_modules/happypack/lib/WebpackUtils.js:122:9)
at resolveLoaders (/__CENSORED__/node_modules/happypack/lib/HappyPlugin.js:137:20)
at /__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:713:13
at iterate (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:262:13)
at async.forEachOfSeries.async.eachOfSeries (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:281:9)
at _parallel (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:712:9)
at Object.async.series (/__CENSORED__/node_modules/happypack/node_modules/async/lib/async.js:734:9)
at HappyPlugin.start (/__CENSORED__/node_modules/happypack/lib/HappyPlugin.js:127:9)
at /__CENSORED__/node_modules/happypack/lib/HappyPlugin.js:100:12
at AsyncSeriesHook.eval [as callAsync] (eval at create (/__CENSORED__/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:7:1)
at AsyncSeriesHook.lazyCompileHook [as _callAsync] (/__CENSORED__/node_modules/tapable/lib/Hook.js:35:21)
at Watching._go (/__CENSORED__/node_modules/webpack/lib/Compiler.js:54:32)
at Watching.compiler.readRecords.err (/__CENSORED__/node_modules/webpack/lib/Compiler.js:46:9)
at Compiler.readRecords (/__CENSORED__/node_modules/webpack/lib/Compiler.js:500:11)
at new Watching (/__CENSORED__/node_modules/webpack/lib/Compiler.js:43:17)
at Compiler.watch (/__CENSORED__/node_modules/webpack/lib/Compiler.js:334:20)
```

https://github.com/amireh/happypack/blob/f153db542a3a97ea27557b1c24e7b11c4722db1b/lib/WebpackUtils.js#L130-L138

Refs https://github.com/webpack/webpack/issues/6132#issuecomment-352570048

**Comment by pksjce** [comment]
Is there a workaround or solution for this?

**Comment by amireh** [comment]
I looked briefly at incorporating sokra's comment on the backward compat layer but didn't get it to work and didn't look further. I'll look again when webpack 4 is released.

**Comment by pksjce** [comment]
I saw the comment too. What does it mean? What is the backward compat layer exactly?

**Comment by sompylasar** [comment]
Compare `webpack@3.10.0`:

https://github.com/webpack/webpack/blob/b545b519b2024e3f8be3041385bd326bf5d24449/lib/Compiler.js#L185-L189

https://github.com/webpack/webpack/blob/215647d9fcf2cfa69c83b3721ae7058ccd31d851/lib/WebpackOptionsApply.js#L283-L292

https://github.com/webpack/enhanced-resolve/blob/49cddd1c5757849b1e0b53b9c765525b840c3b59/lib/ResolverFactory.js#L35

With `webpack@4.0.0-alpha.4`:

https://github.com/webpack/webpack/blob/37cc6f4f6ec178d0406dcf816de690cab9b1bf8f/lib/Compiler.js#L85-L146

**Comment by pksjce** [comment]
Ah! I think I get it. Lemme see if I can send a PR for this.

**Comment by Hoten** [comment]
Hey @pksjce, I'm wondering if you've made any progress on this yet?

**Comment by damianobarbati** [comment]
I'm getting a slightly different error:
```
/Users/damz/Desktop/yarsk/node_modules/happypack/lib/WebpackUtils.js:138
    if (resolve.length === 4) {
                ^

TypeError: Cannot read property 'length' of undefined
    at resolveLoader (/Users/damz/Desktop/yarsk/node_modules/happypack/lib/WebpackUtils.js:138:17)
```

But I guess these are general compatibility errors.

**Comment by wilbert-abreu** [comment]
As webpack 4 comes out of alpha, I'm getting the same error as @damianobarbati.

**Comment by Yelmuf** [comment]
Thanks to author for creating this loader. With the help of other loaders and some refactor we managed to decrease dev warm build time from 707s to ~20s.

Now trying to upgrade webpack to latest version, but getting same error as TS.

Q: Anybody got happypack working with webpack 4 just released? I couldn't, even without cache-loader, and it seems to be caused by some breaking changes.

**Comment by amireh** [comment]
What I'm worried about is that the changes to the plugin architecture (or at least to the official API) prevent happypack from doing what it's been doing since v1, and that is: it tries to resolve the loaders you specify in its config *before* any modules are compiled yet *after* the compilation is initiated. If the resolver is no longer reachable at that stage, I'm not sure how a patch would solve this short of finding a different way to resolve those loaders.

Unfortunately my work right now doesn't allow me to work on this and my time is limited. If anyone's willing to contribute, I'd be happy to help. To provide some context on what the functionality is that got broken:

Back when happypack was in beta you had to provide absolute paths to your loaders and not, for example, `babel-loader`. This was because it wasn't trivial to find out which loader to resolve using just a name since webpack supports custom resolving rules (i.e. `resolveLoader`). So the approach that was taken was to simply use webpack's resolver at an early stage and feed the resolved paths back to happypack's plugin.

TL;DR - to re-iterate, what needs to be done is either of two things:

1. using webpack 4 APIs find out how to resolve loaders after a compilation is initiated but before any module gets processed
2. find a different approach to let people specify their loaders without absolute paths and not rely on webpack for resolving

I'll try to look at this over the weekend but no promises.

**Comment by sompylasar** [comment]
@sokra @TheLarkInn Hi, could you please provide some insights into webpack 4 APIs for this:
> using webpack 4 APIs find out how to resolve loaders after a compilation is initiated but before any module gets processed

Thanks!

**Comment by curiousercreative** [comment]
I began to update HappyPack for webpack 4 starting with the video here: https://medium.com/webpack/webpack-4-migration-guide-for-plugins-loaders-20a79b927202. Unfortunately, I wasn't able to get HappyPack working. Hope this can be helpful to another :)

**Comment by kamiyo** [comment]
Has anyone taken a look at using this to resolve modules?: https://github.com/webpack/enhanced-resolve

I believe this is what the new compiler.ResolverFactory is based on?

I was digging around a couple of nights ago trying to use the new plugin system, but I think I'm a bit over my head with this.

**Comment by marcokam** [comment]
It looks like we can still resolve loaders using different syntax.  I was able to get our happypack build working on both webpack 3 and 4 by making a few changes:
marcokam/happypack@e3fa04b

Something else seems to have changed, as after I fixed the loaders problem there was an issue in HappyLoader.js.  It looks like the `options` property isn't on the loader anymore, so I added a few guards for the webpack 4 case.

Submitted on #215

**Comment by amireh** [comment]
Try out `happypack@next` or `happypack@5.0.0-beta.1` and confirm if it's working in your setup. I just tested it on a project that's recently been upgraded to webpack 4 and seems to work OK.

Thanks to @marcokam for the contribution.

**Comment by Ggpsv** [comment]
Hello, I've recently tried out happypack@next and it is now working with our webpack v4 setup.

**Comment by cvanem** [comment]
happppack@5.0.0-beta.1 worked with Webpack 4 + ts-loader for me.  Unfortunately, I didn't see any speed improvement.  I am running with 6 cores and tried various combinations of workers but it seemd to consistently take about 100-200 ms longer (8600 ms total) than just running ts-loader + fork-ts-checker-webpack-plugin without happypack.

**Comment by jessedvrs** [comment]
Webpack 4 with happypack@5.0.0-beta.1 works for me. From 33838ms to 18724ms 🎉

**Comment by sompylasar** [comment]
`happypack@5.0.0-beta.1` `webpack@4.1.1`

```
DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
  at HappyPlugin.apply (/__CENSORED__/node_modules/happypack/lib/HappyPlugin.js:96:12)
  at webpack (/__CENSORED__/node_modules/webpack/lib/webpack.js:37:12)
  at Promise (/__CENSORED__/webpack/watchDevServer.js:19:22)
  at new Promise (<anonymous>)
  at watchDevServer (/__CENSORED__/webpack/watchDevServer.js:13:10)
  at Promise.all (/__CENSORED__/bin/run-web-frontend.js:64:13)
  at Promise.resolve.then (/__CENSORED__/bin/run-web-frontend.js:75:11)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)
  at Function.Module.runMain (module.js:667:11)
  at startup (bootstrap_node.js:187:16)
  at bootstrap_node.js:608:3
```

https://github.com/amireh/happypack/blob/523f63fdcdb903053cb0876f30c625dbfdedfefa/lib/HappyPlugin.js#L96-L112

**Comment by sompylasar** [comment]
To make it compatible with both `webpack@3` and `webpack@4`, check for `compiler.hooks` and `compilation.hooks`, for example: https://github.com/webpack-contrib/webpack-bundle-analyzer/blob/7c09fa8fbcbea92c165b016989e27291f727a2c3/src/BundleAnalyzerPlugin.js#L63-L67

```js
    if (compiler.hooks) {
      compiler.hooks.done.tap('webpack-bundle-analyzer', done);
    } else {
      compiler.plugin('done', done);
    }
```

**Comment by lerit** [comment]
yes,even update to happypack@5.0.0-beta.1 webpack@4.1.1,still face the problem

**Comment by amireh** [comment]
@lerit what problem? The deprecation notice?

**Comment by lerit** [comment]
@amireh  thank you!i update to happypack@5.0.0-beta.1(webpack v4.2.0)
when use happypack, it throw error from webpack
![untitled](https://user-images.githubusercontent.com/4743467/37865373-9102216c-2fb6-11e8-814a-b906276a3141.png)

**Comment by lerit** [comment]
i read webpack config and complie it as below:
const compiler = webpack(webpackDevConfig)
  compiler.run((err, stats) => {
    //do something
  })
  compiler.watch({}, (err, stats) => {
//do something
  })

**Comment by amireh** [comment]
that's just a deprecation notice not an error, it's been fixed in `5.0.0-beta.2`

**Comment by timidak** [comment]
is this issue fixed? seems like I'm still getting the deprecation warning in `5.0.0-beta.3`

**Comment by amireh** [comment]
Hmm.. I'm not seeing the deprecation notice anymore. Are you sure you've updated? Or perhaps we've got a broken release, I'll double-check.

**Comment by amireh** [comment]
The issue shouldn't be present, I just double-checked. Please verify the deprecations aren't caused by other dependencies you're having. See [this comment](https://github.com/webpack/loader-utils/issues/75#issuecomment-285637755) on how to trace deprecations back to their modules.

**Comment by amireh** [comment]
Version 5.0.0 was just released and should fix this issue so I'll close this one. Please open another if you come across other issues against webpack 4.

Thanks everyone for the help!
